libp2p/autotls/acme/api.nim

```diff
@@ -0,0 +1,395 @@
+      raise newException(ACMEError, "Authorizations field is empty")
+    ACMEChallengeResponse(
+      status: challengeResponseBody.status,
+      authorizations: challengeResponseBody.authorizations,
+      finalize: challengeResponseBody.finalize,
+      orderURL: headers.getString("location"),
+    )
+
+proc requestAuthorizations(
+    self: ACMEApi, authorizations: seq[string], key: KeyPair, kid: Kid
+): Future[ACMEAuthorizationsResponse] {.async: (raises: [ACMEError, CancelledError]).} =
+  handleError("requestAuthorizations"):
+    doAssert authorizations.len > 0
+    let rawResponse =
+      await HttpClientRequestRef.get(self.session, authorizations[0]).get().send()
+    let body = await rawResponse.getResponseBody()
+    body.to(ACMEAuthorizationsResponse)
+
+proc requestChallenge*(
+    self: ACMEApi, domains: seq[string], key: KeyPair, kid: Kid
+): Future[ACMEChallengeResponseWrapper] {.async: (raises: [ACMEError, CancelledError]).} =
+  let challengeResponse = await self.requestNewOrder(domains, key, kid)
+
+  let authorizationsResponse =
+    await self.requestAuthorizations(challengeResponse.authorizations, key, kid)
+
+  return ACMEChallengeResponseWrapper(
+    finalizeURL: challengeResponse.finalize,
+    orderURL: challengeResponse.orderURL,
+    dns01: authorizationsResponse.challenges.filterIt(it.`type` == "dns-01")[0],
+  )
+
+proc requestCheck(
+    self: ACMEApi, checkURL: string, key: KeyPair, kid: Kid
+): Future[ACMECheckResponse] {.async: (raises: [ACMEError, CancelledError]).} =
+  handleError("requestCheck"):
+    let rawResponse =
+      await HttpClientRequestRef.get(self.session, checkURL).get().send()
+    let body = await rawResponse.getResponseBody()
+    let headers = rawResponse.headers
+    let retryAfter =
+      try:
+        parseInt(rawResponse.headers.getString("Retry-After")).seconds
+      except ValueError:
+        DefaultChalCompletedRetryTime
+
+    ACMECheckResponse(status: body["status"].getStr, retryAfter: retryAfter)
+
+proc requestCompleted(
+    self: ACMEApi, chalURL: string, key: KeyPair, kid: Kid
+): Future[ACMECompletedResponse] {.async: (raises: [ACMEError, CancelledError]).} =
+  handleError("challengeCompleted (send notify)"):
+    let payload =
+      await self.createSignedAcmeRequest(chalURL, %*{}, key, kid = Opt.some(kid))
+    let rawResponse = await self.post(chalURL, payload)
+    let body = await rawResponse.getResponseBody()
+    body.to(ACMECompletedResponse)
+
+proc challengeCompleted*(
+    self: ACMEApi,
+    chalURL: string,
+    key: KeyPair,
+    kid: Kid,
+    retries: int = DefaultChalCompletedRetries,
+): Future[void] {.async: (raises: [ACMEError, CancelledError]).} =
+  let completedResponse = await self.requestCompleted(chalURL, key, kid)
+  # check until acme server is done (poll validation)
+  for i in 0 .. retries:
+    let checkResponse = await self.requestCheck(completedResponse.checkURL, key, kid)
+    case checkResponse.status
+    of "pending":
```

Comment:
Since "valid" and "pending" are used too in line 361, consider having an enum, with an utils function like

```nim
type
  ResponseStatus = enum
    pending,
    valid,

proc toResponseStatus(s: string): ResponseStatus =
  try:
    Opt.some(parseEnum[ResponseStatus](s))
  except ValueError:
    Opt.none(ResponseStatus)
```
And then the code would look like
```nim
let checkResponseStatus = checkResponse.status.toResponseStatus.valueOr:
  raise newException(
        ACMEError,
        "Failed challenge completion, got '" & checkResponse.status &
          "'",
      )

case checkResponseStatus:
  of ResponseStatus.pending:
       ........
  of ResponseStatus.valid:
       ........
```

Reply:
If it's an overkill, i guess having some constants should be fine too :)

Reply:
parsing to enum adds safety because if the value is not recognized when parsing it will throw an error. 

if for any reason api changes (status field name changes), or code changes (in a way that's not deserializing correctly) current code will always process this case with the final `else` branch. which may not be incorrect, based on the code usecase, but it is a swallowing an error.